_posts/2018-01-03-Continuous_delivery_with_episerver_dxcs.md

```diff
@@ -0,0 +1,459 @@
+
+## Continuous delivery (CD) and continuous integration (CI)
+
+For sure there are better references about CI and CD than following will be. Actually if you have hard time in convincing your product owner to continuous delivery we have [a blog](http://dev.solita.fi/2016/09/06/CD-for-UX-and-PO.html) for that purpose as well. For me continuous integration means merging work in git repository hourly with coworkers working on the same things. In the other hand continuous delivery means pushing the work to customer as fast as we can. This is fine and simple. People work on the code and then you deployment it once you are confident with it. When you speed up the cycle of your deployments your customer will have new features faster and you they can steer the direction of your development in a truly agile way. Characteristics of CI and CD are below. 
+
+Continuous integration:
+* Make code available to other developers
+* Make sure code builds after merging to works of other people 
+* Track the quality of your code 
+* Test your code 
+
+Continuous delivery:
+* Make code available to your end-users 
+* Make sure that you can deploy any time 
+* Track the quality of your runtime 
+* Test your deployments 
+
+I feel that continuous integration is something that software business is doing already with high adaptation rate. Continuous delivery in the other hand is something that many are still learning. Continuous delivery definitely needs more maturity of the development team since you need to automate a lot of things to be able to say to your CI ecosystem to deploy automatically. You will need a process that makes all kind of sanity checks across the line to trust the deployments on the hand of the automated system. The good thing is once you get on the road it will be a lot easier to adjust the process when you have ability to repeat it multiple times a day. 
+
+The hardest question asked about continuous delivery is most often the following: "How can your customer decide when a feature is good enough for release?". This is something that you need to decide per feature. With Episerver creating new content features is easy. Customer tests the new content types and decides when they serve them well enough to be published on the live. Changing existing content can be done with feature toggles. Of course creating a feature toggle is a change too and can be errornous. Here comes the place the need for automatic testing. You need to catch the errors before they get on the production. The hardest kind of changes are cross-cutting concerns like authentication or database schema changes. You need to be super careful how do you implement and roll this kind of features on the production. In the end the most important thing is that your development organization (developers and business people) have the correct working habits and mindset for failing fast and failing often. Some day you will fail and with CD you will be most likely to recover faster. 
```

Comment:
is often "How can your customer decide when a feature is good enough for release?" laittaisin ehkä jotenkin noin.
serve them well enough to be published.
errornous -> error prone?
From this emerges the need for automatic testing.
You need to catch errors before they get on the production. (Ottaisin the artikkelin pois errorsin edestä)
You need to be super careful in how do you implement and roll these kinds of features to the production.
the mindset